### ISSUE TITLE
What to do when the received message type is not known in advance

### ISSUE THREAD

**Issue by vorner** [jbaublitz/neli#50]
Sometimes, the socket can yield multiple types of messages. Currently, there's not an easy way to handle this situation, as the `recv_nl` and `iter` methods require to know the type in advance. I've hit this in the example added in #48.

One possible way is to use the `recv` call directly and then handle the parsing semi-manually, eg first parse into `Nlhdrmsg<u16, Nlempty>`, and then based on the u16 type decide into what to parse the buffer once again. While possible, this seems inelegant.

I've two ideas what could be done about that.

* Similar to how `Nlattr<_, Vec<u8>>` has methods to parse the payload as something (https://docs.rs/neli/0.4.3/neli/nlattr/struct.Nlattr.html#method.get_payload_as), such a method could be added to `Nlhdrmsg<u16, Vec<u8>>`. The caller could then decide based on the u16 type.
* Some kind of macro that would hide any kind of ugliness for the manual parsing. For example, crossbeam-channel has one such thing for receiving from multiple channels, each one returning potentially different message type, so this should be possible. Then one could write something like:

```rust
neli_recv! ( socket => 
    Err(e) => panic!("Heeeelp"),
    Ok(NetfilterMsg::LogPacket, packet: LogPacket) => dbg!(packet),
    Ok(Nlmsg::Done, _) => return,
)
```

Or is there a better way? Is there a way I've missed in the documentation?

**Comment by skrap** [comment]
I have run into exactly this issue as well, and ended up writing my own code to separate the receiving and parsing phases of the connection.  Since the receiving (essentially the read from the socket) and parsing seem to have separate concerns, it might make sense to separate them.  I would prefer that option.

However, personally I find hiding the details inside a macro to often be counterproductive.  I had to learn a lot about neli's macro system before I was able to use the crate effectively, which cost me more time than it would have to just read the more verbose macro-free version of the code.  The macros also typically defeat IDEs and other tools which might aid code comprehension as well.  I'd rather make users go through the motions of receiving and parsing than hiding everything behind a relatively opaque macro.

**Comment by jbaublitz** [comment]
@skrap I understand the frustration debugging macro magic, and I'll consider that. Sometimes this is the most ergonomic way to handle these sorts of situations simply because of type constraints, but I'll look at both of @vorner's suggestions.

> Similar to how Nlattr<_, Vec<u8>> has methods to parse the payload as something (https://docs.rs/neli/0.4.3/neli/nlattr/struct.Nlattr.html#method.get_payload_as), such a method could be added to Nlhdrmsg<u16, Vec<u8>>. The caller could then decide based on the u16 type.

@vorner Ideally I'd like to provide some level of control either in a macro or a function call that would allow the user to attempt parsing the data as a type - if it fails, rewind and try another. Ultimately I'd like it to dispatch the proper parsing method and return the parsed data. This may be able to be accomplished by both and if I can get that behavior without a macro, I'll probably do that.

I agree that #45 might be the key to this kind of behavior, so I'm going to address that before I do any work on this.

**Comment by jbaublitz** [comment]
Blocked by #45

**Comment by jbaublitz** [comment]
This should be resolved in branch `v0.5.0`. In particular, there is a new method in `recv_nl` which allows the caller to control the behavior of rewinding or skipping on error. This, in addition to new flexibility in constants (see #19) should resolve this issue.
